packages/@aws-cdk/core/lib/lazy.ts

```diff
@@ -150,50 +349,65 @@ abstract class LazyBase implements IResolvable {
 
 }
 
-class LazyString extends LazyBase {
-  constructor(private readonly producer: IStringProducer) {
+class LazyString extends LazyBase<string> {
+
+  constructor(private readonly producer: IStringProducer, private readonly cache: boolean) {
     super();
   }
 
   public resolve(context: IResolveContext) {
-    return this.producer.produce(context);
+    if (this.cache) {
+      return this._cached ?? (this._cached = this.producer.produce(context));
+    } else {
+      return this.producer.produce(context);
+    }
   }
 }
 
-class LazyNumber extends LazyBase {
-  constructor(private readonly producer: INumberProducer) {
+class LazyNumber extends LazyBase<number> {
+  constructor(private readonly producer: INumberProducer, private readonly cache: boolean) {
     super();
   }
 
   public resolve(context: IResolveContext) {
-    return this.producer.produce(context);
+    if (this.cache) {
+      return this._cached ?? (this._cached = this.producer.produce(context));
+    } else {
+      return this.producer.produce(context);
+    }
```

Comment:
Reduce this code duplication. The only additional code will be to special case an Array where you return `undefined` if the length is 0.